File: 024. 两两交换链表中的节点.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def swapPairs(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0) # type: ignore
        dummy.next = head
        pre = dummy
        
        # 只要 pre 后面还有完整的两个节点，就说明可以凑成一对进行交换
        while pre.next and pre.next.next:
            # 在循环内部安全地定义 a 和 b，绝对不会报空指针错误
            a = pre.next
            b = pre.next.next
            
            # 你的核心交换逻辑（一模一样，完美运行）
            a.next = b.next
            b.next = a
            pre.next = b
            
            # 关键修复：为下一轮交换做准备
            # 交换后变成了 pre -> b -> a -> 下一对
            # 所以下一轮的前驱节点应该是 a！
            pre = a
            
        return dummy.next

    # 冗余解法
    def swapPairs2(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0)  # type: ignore
        dummy.next = head
        pre = dummy
        a = dummy.next if dummy else None
        b = a.next if a else None
        while b:
            a.next = b.next
            b.next = a
            pre.next = b
            pre = a
            a = pre.next if pre else None
            b = a.next if a else None
        return dummy.next

File: 024. 两两交换链表中的节点.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	if head.Next == nil {
		return head
	}

	dummy := &ListNode{Next: head}
	p1, p2 := dummy, head

	for {
		p1.Next = p2.Next
		p2.Next = p2.Next.Next
		p1.Next.Next = p2

		if p2.Next == nil || p2.Next.Next == nil {
			break
		}

		p1 = p2
		p2 = p2.Next
	}

	return dummy.Next
}
